sw/device/lib/dif/dif_spi_host_unittest.cc

```diff
@@ -0,0 +1,517 @@
+  uint8_t txbuf[32];
+  uint8_t rxbuf[32];
+  dif_spi_host_segment segments[] = {
+      {
+          .type = kDifSpiHostSegmentTypeBidirectional,
+          .bidir =
+              {
+                  .width = kDifSpiHostWidthStandard,
+                  .txbuf = txbuf,
+                  .rxbuf = rxbuf,
+                  .length = sizeof(txbuf),
+              },
+      },
+  };
+
+  EXPECT_WRITE32(SPI_HOST_CSID_REG_OFFSET, 0);
+  EXPECT_READY(true);
+  EXPECT_CALL(fifo_, write(&spi_host_, txbuf, sizeof(txbuf)));
+  EXPECT_COMMAND_REG(
+      /*length=*/sizeof(txbuf), /*width=*/kDifSpiHostWidthStandard,
+      /*direction=*/kDifSpiHostDirectionBidirectional, /*last=*/true);
+  EXPECT_CALL(fifo_, read(&spi_host_, rxbuf, sizeof(rxbuf)));
+
+  EXPECT_EQ(
+      dif_spi_host_transaction(&spi_host_, 0, segments, ARRAYSIZE(segments)),
+      kDifOk);
+}
+
+// Checks that multiple segments are sent correctly.
+TEST_F(TransactionTest, MultiSegmentTxRx) {
+  uint8_t txbuf[32];
+  uint8_t rxbuf[64];
+  dif_spi_host_segment segments[] = {
+      {
+          .type = kDifSpiHostSegmentTypeTx,
+          .tx =
+              {
+                  .width = kDifSpiHostWidthDual,
+                  .buf = txbuf,
+                  .length = sizeof(txbuf),
+              },
+      },
+      {
+          .type = kDifSpiHostSegmentTypeRx,
+          .rx =
+              {
+                  .width = kDifSpiHostWidthDual,
+                  .buf = rxbuf,
+                  .length = sizeof(rxbuf),
+              },
+      },
+  };
+
+  EXPECT_WRITE32(SPI_HOST_CSID_REG_OFFSET, 0);
+  EXPECT_READY(true);
+  EXPECT_CALL(fifo_, write(&spi_host_, txbuf, sizeof(txbuf)));
+  EXPECT_COMMAND_REG(/*length=*/sizeof(txbuf), /*width=*/kDifSpiHostWidthDual,
+                     /*direction=*/kDifSpiHostDirectionTx, /*last=*/false);
+  EXPECT_READY(true);
+  EXPECT_COMMAND_REG(/*length=*/sizeof(rxbuf), /*width=*/kDifSpiHostWidthDual,
+                     /*direction=*/kDifSpiHostDirectionRx, /*last=*/true);
+  EXPECT_CALL(fifo_, read(&spi_host_, rxbuf, sizeof(rxbuf)));
+
+  EXPECT_EQ(
+      dif_spi_host_transaction(&spi_host_, 0, segments, ARRAYSIZE(segments)),
+      kDifOk);
+}
+
+class FifoTest : public SpiHostTest {};
+
+// Checks that an aligned source buffer is written as 32-bit words into the
+// transmit FIFO.
+TEST_F(FifoTest, AlignedWrite) {
+  uint32_t buffer[] = {1, 2};
+
+  EXPECT_TXQD(0);
+  EXPECT_WRITE32(SPI_HOST_TXDATA_REG_OFFSET, 1);
+  EXPECT_TXQD(0);
+  EXPECT_WRITE32(SPI_HOST_TXDATA_REG_OFFSET, 2);
+
+  dif_spi_host_fifo_write(&spi_host_, buffer, sizeof(buffer));
+}
+
+// Checks that a misaligned source buffer is written as bytes into the
+// transmit FIFO until alignment is reached and then written as 32-bit words.
+TEST_F(FifoTest, MisalignedWrite) {
+  // We'll intentionally mis-align the buffer by 1 when calling
+  // dif_spi_host_fifo_write.
+  uint8_t buffer[] __attribute__((aligned(4))) = {0, 1, 2, 3, 4, 5, 6, 7, 8};
```

Comment:
You can use `alignas` instead of the compiler extension.

Reply:
Done.